src/cart/src/gurt/dlog.c

```diff
@@ -365,6 +405,136 @@ do {								\
 		__func__, __LINE__, err,  ## __VA_ARGS__);	\
 } while (0)
 
+#define LOG_BUF_SIZE	(16 << 10)
+
+/**
+ * This function can do a few things:
+ * - copy log message @msg to log buffer
+ * - if the log buffer is full, write it to log file
+ * - if the log file is too large (exceeds to threshold), rename it to
+ *   original_name.old, and create a new log file.
+ *
+ * If @flush is true, it writes log buffer to log file immediately.
+ */
+static int
+d_log_write(char *msg, int len, bool flush)
```

Comment:
do we have a guarantee somehow that d_log_write() will not end up being called concurrently by multiple threads attempting to do logging? 
Currently in d_vlog we use per-thread local buffers, seeming to indicate that d_vlog can end up being called from multiple threads at the same time.
364         static __thread char b[DLOG_TBSIZ];

If thats the case then stuff like 
"mst.log_buf_nob += len;"  would need to be protected by locks and/or atomic increments/checks

Reply:
I agree with @frostedcmos and in my my opinion there are much more operations here which needs to be protected in a bigger critical section to become indivisible/atomic and thus prevent issues like intermediate buffer/log_buf[] corruption, or even multiple buffer allocs, or duplicated file operations.
And BTW, I also agree that having a per-thread buffer appears like a very simple way to fix some of this issues (the buffer related ones, not the file operations), and this only if it is guaranteed that ULTs will not be preemptable but will only switch context on their own, which hopefully seems to be the case!

Reply:
Well in fact after better reading the whole thing, I think code is already protected against races using mst.clogmux mutex and clog_lock()/clog_unlock() macros usage in d_vlog() !!